Implemented interface states and models

shartte opened the pull request:
Implemented interface model and directed interface arrows.

Introduced a new property type that can hold an AEPartDirection, because only that contains "INTERNAL", which I needed for the omnidirectional interface.

This relies on setting blockstates from tile entites and persists the block state in meta, since its big enough to hold the directional information.

Since the model already rotates correctly based on the point_at property, I disabled the FORWARD/UP properties for the blocks and made the caching rotating model skip such blocks.

This should fix #26

shartte committed:
Implemented interface states and models.

dpeter99 commented on src/main/resources/assets/appliedenergistics2/models/block/tile.BlockInterfaceDirected.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "parent": "block/cube",
```
Couldn't this be a `cube_` something from vanilla predefined parents?

Elix-x commented:
> This relies on setting blockstates from tile entites and persists the block state in meta, since its big enough to hold the directional information.
> 
> Since the model already rotates correctly based on the point_at property, I disabled the FORWARD/UP properties for the blocks and made the caching rotating model skip such blocks.

As we decided with @yueh, **ALL** blocks should be rotatable on all axes. Including ones for which it makes no sense.

yueh commented:
I am neither really sure about using `AEPartLocation` nor a `Property` for these data.

`AEPartLocation` has the semantic of where a part is, not where it is pointing at or even designed for being used with blocks. Also the interface is not pointing at the center or internal, but actually everywhere. It is more like `directional && pointingAtFace`. So it could also use a bool + `EnumFacing`.
In general there are a few important differences between a undirectional and directional interface regardless of the direction it faces.

But the `Property` is the main issue. The interface already serialises a bunch of data, why introduce the facing as exception to it? Also the direction is very frequently used for the tile entity, easily in the range of several thousand calls per tick. While the property should only be used once on a chunk rerender.
Going through the block, propertymap and what not each time is simply not ideal. Also it is needed a few other locations like for the `IGridBlock`. We could cache it. But then we have to deal with multiple sources of truth, correctly invalidating 3 or more caches and so on.
As we already serialise all of the required data (short of being directional or not) through the tile entity, why not use an `IUnlistedProperty` and use the tile entity as SSOT?

Elix-x commented:
@yueh little remark: `IUnlistedProperty` can't be passed to jsons, but a property from `getExtendedState` can (though they don't make much of difference).

yueh commented:
Did forge not announce you can use them in their own json format or something similar?

But it is mostly about serialising the direction as redundant metadata.

Elix-x commented:
> Did forge not announce you can use them in their own json format or something similar?

You can use them in custom `IBakedModels`. Last time i tried with json, it didn't work, but i'll double check.

yueh commented:
Also keep in mind that there is still the feature request for colourable machines. So something like a green north-facing interface will become a thing. (The feature literally needs the rendering part, functionality already exists)

Elix-x commented:
> Also keep in mind that there is still the feature request for colourable machines. So something like a green north-facing interface will become a thing. (The feature literally needs the rendering part, functionality already exists)

You said that just in time...
_So i need to apply the way i recolor cables to machines. Should be straight forward._
@shartte no need to do that part yet.

yueh commented:
Just to be careful, it should only recolour some faces. Like the cable itself, but not a part. Except on a part it should recolour the status light or things like the terminal colour.

But no idea for a "green interface". Recolour the whole thing? One or two of the dark rings? Just the center and maybe the arrows for directional ones? Same also applies to drives, assemblers, io ports and what not. Iirc chests and crafting monitors can already be recoloured, but that only applies to the colour of the monitor itself, not the grid colour.

Elix-x commented:
That's completely up to texturer/modeler to decide. Hence, 100% RP compatible.

shartte commented:
I will revise the PR to use the usual suspects (forward, up) and a boolean
Property to decide omnidirectional=true|false for selecting the Model. I am
sonewhat concerned about the useless duplicate Models we'll be
auto-generating for the omnidirectional variant, but I assume we can
optimize that later.
I do hope we are not gonna make the user cycle through 36 interface states
(6 \* 6) of which 30 are superfluous however. I also only used the
aepartdurection enum, because it was already abused in that way in
TileInterface.

tl;dr will amend PR

On 18 Aug 2016 12:04 p.m., "Elix_x" notifications@github.com wrote:

> That's completely up to texturer/modeler to decide. Hence, 100% RP
> compatible.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/AEModernMCPort/Applied-Energistics-2/pull/66#issuecomment-240680102,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABM_V54nLU2z-56_65VIxIB-CSTLevCcks5qhC4agaJpZM4Jm-02
> .

yueh commented:
The rotations for interfaces always handled a bit different due to the directional/undirectional. It just cycles through 3 states, facing towards the side clicked at, the opposite and then back to undirectional.

I have not really looked at the rotation methods for blockstates. But it would probably a better idea to handle it there, even when internally saving all 24 permutations. Afaik most wrenches in 1.8+ already use them to rotate things. So it can provide a generic way instead of potentially having to support every wrench out there and let it behave correctly.

Elix-x commented:
> Afaik most wrenches in 1.8+ already use them to rotate things. So it can provide a generic way instead of potentially having to support every wrench out there and let it behave correctly.

No, from 1.9, they use rotation methods provided by vanilla, as vanilla never provided even 6 sides facing property (in 1.8, they had to guess property by name and valid values by blind trying).

yueh commented:
I use 1.8+ more as the current minecraft version. This functionality is simply there, but it is not important, if it was added with 1.8.0 , 1.8.9 or 1.9.4. The basic technology has not really changed that much.
Something like `^1.8.0` would probably be more fitting.

I am not even sure, if we should target 1.10 or 1.11. Minecon is just a month away and we might already see a pre release with it. 1.11 will also have a few breaking changes. For example it finally gets rid of all public fields. Finally everthing is access through setters and getters. Like `ItemStack.increaseStackSizeBy(42)` or so. Or the item field within `ItemStack` will finally be final.

Elix-x commented:
Forge won't come until 1.1 is released. And mojang probably won't release 1.11 right away on minecon, so it will come soon, but not very soon.

Speaking of breaking changes, we already know that they enforced lower casing and changed a bunch of stuff in entities (at least subtypes). Luckily, second one does not affect us.

shartte committed:
Removed pointAt property from Tile Interface and switched to just a flag that indicates the omnidirectionality of the interface since forward/up do the job of pointing it somewhere just fine.

shartte commented:
Okay I amended the PR and touched up TileInterface more than the initial PR was intending to do.

I've removed pointAt from TileInterface and replaced with a simple "omniDirectional" toggle. The interface will now use getUp() for the direction it's pointing and omniDirectional as a toggle between the two JSON models. Rotation is fully handled by the existing mechanism now.

I also added the update packets that were missing so that the omniDirectional state is correctly updated to clients.

Elix-x commented:
`get`**`Forward`** is the one you should use, not `getUp` :wink: .

Also, all we need is sync TE data, this is a work for another day as requires of touching far not only interface (so syncing can be left for now).

shartte commented:
@elix-x Why delay the syncing? It works! And at least you can _test_ this stuff now. Otherwise you never see the omnidirectional state again.

Regarding up/foward: It has to be up because in "resting" state that's how the oriented model is designed. And there's no functional difference with regards to rotation whether its up or forward.

Elix-x commented:
Well, while we accept temporary solutions, they all have to be cleaned up after. It's ok, but it's you again (who will clean up).

> because in "resting" state that's how the oriented model is designed

What? Please elaborate.

> And there's no functional difference with regards to rotation whether its up or forward.

Yes, but, forward is called forward for a reason after all.

yueh commented on src/main/java/appeng/block/misc/BlockInterface.java:
```diff
@@ -52,6 +57,26 @@ public BlockInterface()
 	@Override
+	protected IProperty[] getAEStates()
+	{
+		return new IProperty[] { AE_BLOCK_FORWARD, AE_BLOCK_UP, OMNIDIRECTIONAL };
```
Should that not better call super?

shartte replied:
Yeah, it's a bit annoying that the super method is an array though.

Elix-x replied:
We can change that to `List`, but it won't help in any way.

yueh replied:
Should something like `Stream.concat(super.getAEStates(), new IProperty[] { OMNIDIRECTIONAL }).toArray(IProperty[]::new);` not work?

Elix-x replied:
But do we actually need it?

shartte replied:
What I did here is consistent with all existing overrides of getAEStates. So I'd like to leave this as is and change it in a broader change that also allows for extended block state to be more neatly specified.

yueh replied:
Good question. That might also be something related to synchronizing and the package structure in general. The interface could easily be a subclass of `AEBaseNetworkTileEntity`, which handles commonly shared states like `isActive`, `isPowered` and so on. You do not want to touch every single subclass, just because `AEBaseNetworkTileEntity` got a new property for some reasons.

shartte replied:
Well we're talking block states here. If you extend the property container, you'll most likely have to extend the block variants as well unless you also touch the state mapper. 
In that cleanup we might also come to the conclusion that FORWARD/UP should be unlisted properties anyway and remove our state mapper that currently hides them. Otherwise it's always a bit unclear which properties actually affect the variant-space of a block and which don't. When I started contributing this was one of the things that confused me: That there were no variants required for all forward/up combinations, since I didn't know about the custom state mapper.

shartte commented:
Well by default the oriented model points "up", but I'll change it and see if I can make it use getFoward(). I'd also prefer that, actually.

shartte committed:
Changed interface to use forward direction and not the up property to determine where it is oriented.

shartte commented:
I amended the PR to use getForward() instead of getUp(). Luckily, all I needed to do was rotate the model around the X axis to make it work.

dpeter99 commented:
Ok. But shouldn't we use the new naming for things ?

Elix-x commented:
@dpeter99 it got auto formatted in https://github.com/AEModernMCPort/Applied-Energistics-2/commit/ad4580d34e9e2565c77c4f2a787f5d58b0de078a.
